crates/extensions/c8y_mapper_ext/src/tests.rs

```diff
@@ -2610,6 +2351,275 @@ async fn mapper_processes_operations_concurrently() {
+
+    spawn_dummy_c8y_http_proxy(http);
+
+    // Complete sync phase so that alarm mapping starts
+    trigger_timeout(&mut timer).await;
+
+    let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+    skip_init_messages(&mut mqtt).await;
+
+    // simulate many successful operations that needs to be handled by the mapper
+    mqtt.send(MqttMessage::new(
+            &Topic::new_unchecked("te/device/main///cmd/log_upload/c8y-mapper-1"),
+            json!({
+            "status": "successful",
+            "tedgeUrl": "http://localhost:8888/tedge/file-transfer/test-device/log_upload/c8y-mapper-1",
+            "type": "mosquitto",
+            "dateFrom": "2023-11-28T16:33:50+01:00",
+            "dateTo": "2023-11-29T16:33:50+01:00",
+            "searchText": "ERROR",
+            "lines": 1000
+        })
+                .to_string(),
+        ))
+            .await.unwrap();
+
+    let (download_id, download_request) = dl
+        .recv()
+        .await
+        .expect("DownloadRequest for log_upload should be sent");
+    assert_eq!(download_id, "c8y-mapper-1");
+    assert_eq!(
+        download_request.url,
+        "http://localhost:8888/tedge/file-transfer/test-device/log_upload/c8y-mapper-1"
+    );
+
+    // here it would be good to assert that upload message hasn't been sent yet, but due to the
+    // behaviour of message channels it can't be easily done
+
+    dl.send((
+        "c8y-mapper-1".to_string(),
+        Ok(DownloadResponse {
+            url: "http://localhost:8888/tedge/file-transfer/test-device/log_upload/c8y-mapper-1"
+                .to_string(),
+            file_path: "whatever".into(),
+        }),
+    ))
+    .await
+    .unwrap();
+
+    let (upload_id, _) = ul
+        .recv()
+        .await
+        .expect("UploadRequest for log_upload should be sent");
+    assert_eq!(upload_id, "c8y-mapper-1");
+
+    // now that an upload is ongoing, check that downloads can also be triggered
+    mqtt.send(MqttMessage::new(
+            &Topic::new_unchecked("te/device/main///cmd/config_snapshot/c8y-mapper-2"),
+            json!({
+            "status": "successful",
+            "tedgeUrl": "http://localhost:8888/tedge/file-transfer/test-device/config_snapshot/c8y-mapper-2",
+            "type": "typeA",
+        })
+                .to_string(),
+        ))
+            .await.unwrap();
+
+    let (config_snapshot_id, _) = dl
+        .recv()
+        .await
+        .expect("DownloadRequest for config snapshot should be sent");
+    assert_eq!(config_snapshot_id, "c8y-mapper-2");
+
+    // while download and upload are ongoing, try some other operation that doesn't do download or upload
+    mqtt.send(MqttMessage::new(
+        &Topic::new_unchecked("te/device/main///cmd/restart/c8y-mapper-3"),
+        json!({
+            "status": "successful",
+        })
+        .to_string(),
+    ))
+    .await
+    .unwrap();
+
+    assert_received_contains_str(&mut mqtt, [("c8y/s/us", "503,c8y_Restart")]).await;
+}
```

Comment:
@albinsuresh My bad, by mistake I included the test you were proposing (https://github.com/thin-edge/thin-edge.io/pull/2904#discussion_r1620205761) in this PR, but since it's already here, you can take a look.

Reply:
Looks fine.